x/shield/keeper/proposal.go

```diff
@@ -4,16 +4,182 @@ import (
+	proportion := lossAmt.ToDec().Quo(totalCollateral.ToDec())
+	remaining := lossAmt
+	for i := range providers {
+		var lockAmt sdk.Int
+		if i < len(providers)-1 {
+			lockAmt = providers[i].Collateral.ToDec().Mul(proportion).TruncateInt()
+			if lockAmt.LT(providers[i].Collateral) && lockAmt.LT(remaining) {
+				lockAmt = lockAmt.Add(sdk.OneInt())
+			}
+			remaining = remaining.Sub(lockAmt)
+		} else {
+			lockAmt = remaining
+		}
+		k.LockProvider(ctx, providers[i], types.NewLockedCollateral(proposalID, lockAmt), lockPeriod)
+	}
+
+	// Update shield amount and delete time of the purchase.
+	purchaseList, found := k.GetPurchaseList(ctx, poolID, purchaser)
+	if !found {
+		return types.ErrPurchaseNotFound
+	}
+	var index int
+	for i, entry := range purchaseList.Entries {
+		if entry.PurchaseID == purchaseID {
+			index = i
+			break
+		}
+	}
+	purchase := &purchaseList.Entries[index]
+	if lossAmt.GT(purchase.Shield) {
+		return types.ErrNotEnoughShield
+	}
+	k.DequeuePurchase(ctx, purchaseList, purchase.DeletionTime)
+	purchase.Shield = purchase.Shield.Sub(lossAmt)
+	votingEndTime := ctx.BlockTime().Add(lockPeriod)
+	if purchase.DeletionTime.Before(votingEndTime) {
+		// TODO: correctly update delete time & protection end time
+		purchase.DeletionTime = votingEndTime // temp
+	}
+	k.SetPurchaseList(ctx, purchaseList)
+	k.InsertExpiringPurchaseQueue(ctx, purchaseList, purchase.DeletionTime)
+
+	// Update pool and global pool states
+	pool.Shield = pool.Shield.Sub(lossAmt)
+	k.SetPool(ctx, pool)
+
+	totalLocked = totalLocked.Add(lossAmt)
+	totalCollateral = totalCollateral.Sub(lossAmt)
+	totalShield := k.GetTotalShield(ctx)
+	totalShield = totalShield.Sub(lossAmt)
+	k.SetTotalLocked(ctx, totalLocked)
+	k.SetTotalCollateral(ctx, totalCollateral)
+	k.SetTotalShield(ctx, totalShield)
+
 	return nil
 }
 
+// LockProvider locks the specified amount of collaterals from a provider.
+// If necessary, it extends withdrawing collaterals and, if exist, their
+// linked unbondings as well.
+func (k Keeper) LockProvider(ctx sdk.Context, provider types.Provider, newLock types.LockedCollateral, lockPeriod time.Duration) {
+	originalCollateral := provider.Collateral
+	provider.Locked = provider.Locked.Add(newLock.Amount)
+	provider.Collateral = provider.Collateral.Sub(newLock.Amount)
+	provider.LockedCollaterals = append(provider.LockedCollaterals, newLock)
+
+	// If there are enough bonded delegations backing
+	// locked collaterals, we are done.
+	if provider.DelegationBonded.GTE(provider.Locked) {
+		k.SetProvider(ctx, provider.Address, provider)
+		return
+	}
+
+	// Lenient check:
+	// Consider the amount of non-withdrawing collateral.
+	if provider.Locked.GT(provider.Collateral.Sub(provider.Withdrawing)) {
+		// Stricter check:
+		// Consider the amount of non-withdrawing and withdrawing
+		// collaterals 4 days from now.
+		endTime := ctx.BlockTime().Add(lockPeriod)
+		impendingWithdrawAmount := k.ComputeWithdrawAmountByTime(ctx, endTime)
```

Comment:
`ComputeWithdrawAmountByTime()` returns withdrawing for all providers. Should we consider withdrawing for this provider rather than all providers?

Reply:
Yup. Thank you